pkg/runtime/processor/components.go

```diff
@@ -144,13 +144,13 @@ func (p *Processor) AddPendingComponent(ctx context.Context, comp componentsapi.
 		return false
 	}
 
-	p.pendingComponentsWaiting.Add(1)
+	p.pendingComponentsWaiting.RLock()
+	defer p.pendingComponentsWaiting.RUnlock()
```

Comment:
This seems to be actually changing the logic. Previously, we would only mark as done once the component has been processed. Now we are marking as done at the start of being processed. Please return `p.pendingComponentsWaiting.RUnlock()` to the ctx and close select statements below, as well as line 180 which has been removed below.